### ISSUE TITLE
Make some of internal API public so that other crates can build on top of mockall

### ISSUE THREAD

**Issue by taminomara** [asomers/mockall#318]
TLDR: I'm writing a crate that will allow mocking ethereum smart contracts, and I want it to work nicely with `mockall`. For that, I'd like to request making `SeqHandle` a part of official public API.

Longer explanation.

So, ethereum contracts are like structs (with data and methods) that live in a blockchain. I'm writing a crate that will allow mocking them without using an actual chain. The idea is that you create a mock blockchain and deploy mock contracts onto it. When you deploy a contract, you get an object for setting up expectations. Since we use `mockall` quite extensively in our projects, I've decided to mirror its API in this object:

```rust
let chain = mock_chain();

let contract = chain.deploy_contract(...);
contract.expect_call("MethodA")
    .once()
    .with(...)
    .returning(...);
```

Now I want to implement `in_sequence`, and I want it to work with `mockall`'s native `SeqHandle`. This will allow creating sequences that combine calls to contracts and to normal mocked traits and functions.

**Comment by asomers** [comment]
Can you give an example of how you would use `SeqHandle`?

**Comment by taminomara** [comment]
I suppose I'd use it the same way it is used by the generated code. I.e., my struct that represents mocked contract will have `in_sequence` method, which takes a `Sequence`, generates a `SeqHandle` and stores it. Then, whenever and expectation is satisfied or verified, I'll call `satisfy` or `verify` method on the stored handle.

**Comment by asomers** [comment]
To help me understand, could you provide a code sample, even one that doesn't compile?

**Comment by taminomara** [comment]
Sorry, was away for some time.

Here's an example of our crate's API when used with sequences: https://github.com/gnosis/ethcontract-rs/blob/main/ethcontract-mock/src/test/mod.rs#L118

Specifically, we save mockall's sequence [in our code][1], then we [verify and satisfy it][2] when contract methods are called.

[1]: https://github.com/gnosis/ethcontract-rs/blob/main/ethcontract-mock/src/details/mod.rs#L868
[2]: https://github.com/gnosis/ethcontract-rs/blob/main/ethcontract-mock/src/details/mod.rs#L923-L929

**Comment by asomers** [comment]
It looks like you've reimplemented almost all of Mockall, keeping only `Sequence`, one of the smallest components.  Correct?  If so, I don't think  that's sufficient reason to publicize Sequence's internal API.  If I do that, then I would be committed to keeping it stable.

**Comment by taminomara** [comment]
I mean, not exactly. I've manually implemented the same code for smart contracts that mockall would generate for traits and structs. Thus, it uses this semi-public API for generated code. And I'd like it to be interoperable with mockall's `Sequence`. Otherwise, I'll have to roll my own `Sequence` and loose the entire point of re-implementing mockall's interface.

**Comment by asomers** [comment]
What _is_ the point of reimplementing all of Mockall's generated code, anyway?  Are you trying to mix calls to the same Sequence objects from both generated code and your own custom code?

**Comment by taminomara** [comment]
Yes, that's exactly what I'm doing.
